### PR TITLE
Added BulletPhysics APIs to return internal objects.

### PR DESCRIPTION
**Pull request by AndreaCatania**
Added few APIs, that returns objects pointers, in BulletPhysics.

These give a deep control to modules that uses the physics server directly.

### CODE REVIEW

File: modules/bullet/bullet_physics_server.cpp [hunk 1]
```diff
@@ -1587,5 +1597,10 @@ RigidCollisionObjectBullet *BulletPhysicsServer3D::get_rigid_collisin_object(RID
...
 		return area_owner.getornull(p_object);
 	}
-	return nullptr;
+	ERR_FAIL_V(nullptr);
 }
+
+JointBullet *BulletPhysicsServer3D::get_joint(RID p_rid) const {
+	ERR_FAIL_COND_V(joint_owner.owns(p_rid) == false, nullptr);
+	return joint_owner.getornull(p_rid);
+}
```
Comment (akien-mga) [inline modules/bullet/bullet_physics_server.cpp:1606]: Missing line at EOF.

File: modules/bullet/bullet_physics_server.h [hunk 2]
```diff
@@ -405,11 +405,11 @@ class BulletPhysicsServer3D : public PhysicsServer3D {
 
 	virtual int get_process_info(ProcessInfo p_info);
 
+	SpaceBullet *get_space(RID p_rid) const;
+	ShapeBullet *get_shape(RID p_rid) const;
 	CollisionObjectBullet *get_collisin_object(RID p_object) const;
 	RigidCollisionObjectBullet *get_rigid_collisin_object(RID p_object) const;
```
Comment (akien-mga) [inline modules/bullet/bullet_physics_server.h:411]: "collision" is typo'ed in these methods.

Comment (AndreaCatania) [reply modules/bullet/bullet_physics_server.h]: Fixed

File: modules/bullet/bullet_physics_server.cpp (hunk 1 above)
Comment (AndreaCatania) [reply modules/bullet/bullet_physics_server.cpp]: Fixed

**Review by madmiraal — COMMENTED** [review]
Aside from providing descriptive error messages, the code looks good. However, my questions are more around the motivation: Why is this needed? How does it support Godot as a game engine? I can't find a feature proposal supporting this change. Although, I can imagine people using Godot as a Framework for easy access to Bullet physics wanting greater access to Bullet physics, why is this needed for making games? What modules are people creating that need this?

**Comment by AndreaCatania** [comment]
The people may want to use the specific physics engine directly, so to access the not exposed features or to avoid the `PhysicsServer3D` overhead:
```cpp
BulletPhysicsServer3D* singleton = static_cast<BulletPhysicsServer3D*>(PhysicsServer::get_singleton());
ShapeBullet* bt_shape = singleton->get_shape(shape_rid);
```

This is something already allowed by the engine from a lot of time (for this reason you can't find a proposal), any extra feature too fancy to be exposed as a godot feature, can be used as I showed above.

File: modules/bullet/bullet_physics_server.cpp [hunk 3]
```diff
@@ -1567,7 +1567,17 @@ int BulletPhysicsServer3D::get_process_info(ProcessInfo p_info) {
 	return 0;
 }
 
-CollisionObjectBullet *BulletPhysicsServer3D::get_collisin_object(RID p_object) const {
+SpaceBullet *BulletPhysicsServer3D::get_space(RID p_rid) const {
+	ERR_FAIL_COND_V(space_owner.owns(p_rid) == false, nullptr);
+	return space_owner.getornull(p_rid);
```
[the remark this answers is not in the document]
Comment (AndreaCatania) [reply modules/bullet/bullet_physics_server.cpp:1572]: Done

File: modules/bullet/bullet_physics_server.cpp [hunk 4]
```diff
@@ -1567,7 +1567,17 @@ int BulletPhysicsServer3D::get_process_info(ProcessInfo p_info) {
...
+}
+
+ShapeBullet *BulletPhysicsServer3D::get_shape(RID p_rid) const {
+	ERR_FAIL_COND_V(shape_owner.owns(p_rid) == false, nullptr);
+	return shape_owner.getornull(p_rid);
```
[the remark this answers is not in the document]
Comment (AndreaCatania) [reply modules/bullet/bullet_physics_server.cpp:1577]: Done

File: modules/bullet/bullet_physics_server.cpp [hunk 5]
```diff
@@ -1577,15 +1587,20 @@ CollisionObjectBullet *BulletPhysicsServer3D::get_collisin_object(RID p_object)
 	if (soft_body_owner.owns(p_object)) {
 		return soft_body_owner.getornull(p_object);
 	}
-	return nullptr;
+	ERR_FAIL_V(nullptr);
 }
```
[the remark this answers is not in the document]
Comment (AndreaCatania) [reply modules/bullet/bullet_physics_server.cpp:1591]: Done

File: modules/bullet/bullet_physics_server.cpp [hunk 6]
```diff
@@ -1577,15 +1587,20 @@ CollisionObjectBullet *BulletPhysicsServer3D::get_collisin_object(RID p_object)
...
 	if (area_owner.owns(p_object)) {
 		return area_owner.getornull(p_object);
 	}
-	return nullptr;
+	ERR_FAIL_V(nullptr);
+}
+
+JointBullet *BulletPhysicsServer3D::get_joint(RID p_rid) const {
+	ERR_FAIL_COND_V(joint_owner.owns(p_rid) == false, nullptr);
+	return joint_owner.getornull(p_rid);
```
[the remark this answers is not in the document]
Comment (AndreaCatania) [reply modules/bullet/bullet_physics_server.cpp:1605]: Done

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**